src/ring_theory/non_zero_divisors.lean

```diff
@@ -73,4 +73,28 @@ lemma map_le_non_zero_divisors_of_injective {B : Type*} [integral_domain B]
 le_non_zero_divisors_of_domain (λ h, let ⟨x, hx, hx0⟩ := h in
   zero_ne_one (hM (hf (trans hx0 (f.map_zero.symm)) ▸ hx : 0 ∈ M) 1 (mul_zero 1)).symm)
 
+lemma prod_zero_iff_exists_zero {R : Type*} [comm_semiring R] [no_zero_divisors R] [nontrivial R]
+  {s : multiset R} : s.prod = 0 ↔ ∃ (r : R) (hr : r ∈ s), r = 0 :=
+begin
+split, swap,
+  { rintros ⟨r, hrs, hrz⟩,
+    apply multiset.prod_eq_zero,
+    rwa hrz at hrs },
+  refine multiset.induction _ (λ a s ih, _) s,
+  { rw multiset.prod_zero,
+    intro habs,
+    by_contra, from ne.elim (@one_ne_zero _ _ _) habs },
+  { rw multiset.prod_cons,
+    intro hprod,
+    replace hprod := eq_zero_or_eq_zero_of_mul_eq_zero hprod,
+    cases hprod with ha,
+    { exact ⟨a, multiset.mem_cons_self a s, ha⟩ },
+    { apply (ih hprod).imp _,
+      rintros b ⟨hb₁, hb₂⟩,
+      have this : ∃ (hr : b ∈ a ::ₘ s), b = 0,
```

Comment:
No need to explicitly name something `this`. Just `have : ...` will default to the name `this`. If you're going to give it a name by hand, it might as well be either shorter or more helpful. :-)